grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java

```
@@ -0,0 +1,128 @@
+
+  @Value.Immutable
+  interface RequestInfo {
+
+    String token();
+
+    Map<String, String> headers();
+
+    @Nullable
+    SocketAddress remoteAddress();
+  }
+
+  public NewConnectionInterceptor(
+      Persistence persistence, AuthenticationService authenticationService) {
+    this.persistence = persistence;
+    this.authenticationService = authenticationService;
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    try {
+      String token = headers.get(TOKEN_KEY);
+      if (token == null) {
+        call.close(Status.UNAUTHENTICATED.withDescription("No token provided"), new Metadata());
+        return new NopListener<>();
+      }
+
+      RequestInfo info =
+          ImmutableRequestInfo.builder()
+              .token(token)
+              .headers(convertAndFilterHeaders(headers))
+              .remoteAddress(call.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR))
+              .build();
+
+      Connection connection = connectionCache.get(info);
+
+      Context context = Context.current();
+      context = context.withValue(GrpcService.CONNECTION_KEY, connection);
+      return Contexts.interceptCall(context, call, headers, next);
+    } catch (Exception e) {
+      call.close(
+          Status.UNAUTHENTICATED.withDescription("Invalid token").withCause(e), new Metadata());
+    }
+    return new NopListener<>();
+  }
+
+  private Connection newConnection(RequestInfo info) throws UnauthorizedException {
+    AuthenticationSubject authenticationSubject = authenticationService.validateToken(info.token());
+
+    AuthenticatedUser user = authenticationSubject.asUser();
+    Connection connection;
+    if (!user.isFromExternalAuth()) {
+      SocketAddress remoteAddress = info.remoteAddress();
+      InetSocketAddress inetSocketAddress = DUMMY_ADDRESS;
```

what's the deal with `DUMMY_ADDRESS` :smile: ?

---

This is a best effort attempt to put the correct remote address in the `ClientState` object. If the remote address is not an `InetSocketAddress` I just put a dummy value in there instead of erroring out for something that's not critical. It will almost never not be a `InetSocketAddress` though.